packages/react-menu/src/components/Menu/useMenu.tsx

```diff
@@ -108,61 +100,118 @@ const useMenuSelectableState = (state: MenuState) => {
 };
 
 const useMenuOpenState = (state: MenuState) => {
-  const shouldHandleKeyboadRef = React.useRef(false);
+  const { targetDocument } = useFluent();
   const parentSetOpen = useMenuContext(context => context.setOpen);
   const onOpenChange: MenuState['onOpenChange'] = useEventCallback((e, data) => state.onOpenChange?.(e, data));
 
+  const shouldHandleKeyboadRef = React.useRef(false);
+  const shouldHandleTabRef = React.useRef(false);
+  const pressedShiftRef = React.useRef(false);
+  const setOpenTimeout = React.useRef(0);
+  const enteringTriggerRef = React.useRef(false);
+
   const [open, setOpen] = useControllableValue(state.open, state.defaultOpen);
   state.open = open !== undefined ? open : state.open;
+  const trySetOpen = useEventCallback((e: MenuOpenEvents, data: MenuOpenChangeData) => {
+    const event = e instanceof CustomEvent && e.type === MENU_ENTER_EVENT ? e.detail.nativeEvent : e;
+    onOpenChange?.(event, { ...data });
+
+    if (data.keyboard) {
+      shouldHandleKeyboadRef.current = true;
+      shouldHandleTabRef.current = (e as React.KeyboardEvent).key === 'Tab';
+      pressedShiftRef.current = (e as React.KeyboardEvent).shiftKey;
+    }
+
+    if (data.bubble) {
+      parentSetOpen(e, { ...data });
+    }
+
+    setOpen(data.open);
+  });
 
-  state.setOpen = React.useCallback(
-    (e, data) => {
-      setOpen(prevOpen => {
-        // More than one event (mouse, focus, keyboard) can request the popup to close
-        // We assume the first event is the correct one
-        if (prevOpen !== data.open) {
-          onOpenChange?.(e, { ...data });
-        }
-
-        if (data.keyboard) {
-          shouldHandleKeyboadRef.current = true;
-        }
-
-        if (data.bubble) {
-          parentSetOpen(e, { ...data });
-        }
-
-        return data.open;
-      });
+  state.setOpen = useEventCallback((e: MenuOpenEvents, data: MenuOpenChangeData) => {
+    clearTimeout(setOpenTimeout.current);
+    if (e.type === 'mouseleave' || e.type === 'mouseenter') {
+      if (state.triggerRef.current?.contains(e.target as HTMLElement)) {
+        enteringTriggerRef.current = e.type === 'mouseenter';
+      }
+      // TODO #18315 setTimeout for now to make it easier to consistently call onOpenChange once for each state change
+      setOpenTimeout.current = setTimeout(() => trySetOpen(e, data));
+    } else {
+      trySetOpen(e, data);
+    }
+  });
+
+  useOnClickOutside({
+    contains: elementContains,
+    disabled: !state.open,
+    element: targetDocument,
+    refs: [state.menuPopoverRef, state.triggerRef],
+    callback: e => state.setOpen(e, { open: false }),
+  });
+  useOnMenuMouseEnter({
+    element: document,
```

Comment:
I don't think `document` is correct. What about mw scenarios?

Reply:
ah sorry need to do `useFluent` here

Reply:
fixed